### DISCUSSION TITLE
Best Practice: Starlight living as a proxied subdirectory

### DISCUSSION THREAD

**Discussion by brob** [withastro/starlight#1763] [Q&A] [answered]
I'll be moving a Starlight implementation from its own domain to be proxied under a main domain (from a Nextjs site). 

I'm hoping to get some insight into best practice around links and image URLs and keeping them easy to maintain and run locally from the Starlight repository, but also have them work from the proxy.

The proxy is maindomain.com/docs/ -> Starlight

All links when run under the proxy need to be at `/docs` (images included), but when run as an individual starlight project, they need to be local to the Starlight implementation (as far as I can tell). 

Would love any insights or best practices

**Comment by delucis** [comment, marked as answer]
Hi @brob — good question!

I think for a case like this I would recommend setting Astro’s [`base` config option](https://docs.astro.build/en/reference/configuration-reference/#base) to the proxied path, so something like:

```js
import { defineConfig } from 'astro/config';
import starlight from '@astrojs/starlight';

export default defineConfig({
  base: '/docs/',
  integrations: [
    starlight({
      title: 'My delightful docs site',
    }),
  ],
});
```

This will apply in both cases, so when running the local project everything will also be prefixed with `/docs/`. I wasn’t clear from the description if that would be an issue. If it’s only for local development, I guess it wouldn’t be an issue if you need to navigate to `/docs/` instead of `/` to see the site.

`base` will take care of changing how a bunch of stuff is referenced:

- All assets like JS, CSS etc. For example, a CSS file might be linked to as `/docs/_astro/something.css`
- Starlight’s sidebar will include the base for you in its links.

What Astro/Starlight _don’t_ take care of automatically are links inside your content. For example, if you have some Markdown like this:

```md
Check out [our deployment guide](/deploy/) for more details.
```

We can’t _really_ safely assume you mean `/docs/deploy/` — maybe you really do want to link out below the base? So in this set up, you do kind of ideally want to author those with the `/docs/` prefix.

That said, given you’re moving an existing code base that could be a bit of a pain, so one option could be to add a small remark plugin which processes all links and injects the base automatically.

---

**Edit:** Ah, and one thing that sometimes trips people up: `base` does not create that directory in build output, i.e. `index.html` will still be built to `dist/index.html` not to `dist/docs/index.html`. This means we can support hosts like GitHub Pages that add that prefix automatically. So if that’s an issue, you can do stuff like also setting [`outDir`](https://docs.astro.build/en/reference/configuration-reference/#outdir) to `dist/docs/` in case you need that directory in the deployment.

**Reply by delucis** [reply]
While we’re here, dusted off an old remark plugin I had for the link prefixing part in case it is helpful:

1. Add the required dependencies:

	```sh
	npm i mdast unist-util-visit
	```

2. Save the plugin somewhere in your project (this is TypeScript):

	```ts
	import type { Definition, Link, Root } from 'mdast';
	import { visit } from 'unist-util-visit';

	/** Remark plugin to prepend a path in front of relative link hrefs. */
	export function prefixLinks({ base }: { base: string }) {
	  // Make sure the base path has a trailing slash.
	  if (!base.endsWith('/')) base = base + '/';
	
	  return function remarkPlugin() {
	    return function transform(tree: Root) {
	      function visitor(node: Link | Definition) {
	        // Ignore links that aren’t relative to the root `/`
	        if (!node.url.startsWith('/')) return;
	        // Sanitize URL by removing leading `/`
	        const relativeUrl = node.url.replace(/^.?\//, '');
	        // Prefix with base path
	        node.url = base + relativeUrl;
	      }
	      // Apply our visitor to Markdown links, including definition links
	      visit(tree, 'link', visitor);
	      visit(tree, 'definition', visitor);
	      // Also prefix base to links written with HTML syntax (<a>)
	      visit(tree, 'html', function htmlVisitor(node) {
	        node.value = node.value.replace(
	          /(?<=href=")(?!https?:\/\/)\/?(.+)(?=")/g,
	          `${base}$1`
	        );
	      });
	    };
	  };
	}
	```
	
3. Add the plugin to your `astro.config.mjs`:

	```js
	import { defineConfig } from 'astro/config';
	import starlight from '@astrojs/starlight';
	import { prefixLinks } from './wherever/you/saved/the-plugin.ts';
	
	export default defineConfig({
	  base: '/docs/',
	  markdown: {
	    remarkPlugins: [prefixLinks({ base: '/docs/' })],
	  },
	  integrations: [
	    starlight({
	      title: 'My delightful docs site',
	    }),
	  ],
	});

Link to tiny example Astro project with this in place: https://stackblitz.com/edit/github-x8sahr?file=astro.config.ts

**Reply by brob** [reply]
Thanks for the in-depth answer!

I'd been looking at `base` as a potential solution, so I'm glad that's the right way to go.

My current image strategy in the full docs will need to change, but we can probably do that.


I went to do a test case on this and when I deployed an almost completely out of the box Startlight installation to Vercel with just the `base` added (mostly), it breaks in interesting ways.

![image](https://github.com/withastro/starlight/assets/799360/fa3d3b2b-bcb5-4819-a1de-d2f82588c804)

None of the pages are at `/docs` including the generated CSS (though the CSS being linked to from Starlight is respecting the base... so the CSS file being linked is `/docs/_astro/index.css` but it exists at `/_astro/index.css`. So no CSS is being rendered

It can be tested here: https://starlight-proxy-test.vercel.app/ and the test case repo is [here](https://github.com/brob/starlight-proxy-test). This is the direct Starlight install. It also works when it has no `base` set: i've got a [deploy preview here](https://starlight-proxy-test-git-withoutbase-hygraph.vercel.app/) without the base. I hadn't gotten to putting this in my Next proxy yet

I think I ran into this earlier but assumed it was my proxy breaking it, but it might be that there are some deeper issues with `base`?

**Reply by TheOtterlord** [reply]
Ahh, it's building the output to `/dist/`, assuming the host will serve the files from `/docs/`. If your deploy host serves from the root, you can modify the [`outDir`](https://docs.astro.build/en/reference/configuration-reference/#outdir) to e.g. `./dist/docs` to place everything at the correct level.

```ts
export default defineConfig({
	base: '/docs/',
	outDir: './dist/docs/',
	...
});
```

**Reply by brob** [reply]
Thanks for all your help! I successfully have my Proof of Concept on this and hopefully (barring issues with SEO concerns), I'll be taking a site live later this year with this concept!

Do you think this is a common enough pattern that a blog post/video tutorial on doing it would be beneficial? I may make one

**Reply by delucis** [reply]
Definitely! Always love to see people explain how they tackled these more complex, but not so uncommon scenarios.

**Reply by brob** [reply]
Just for future folks here's my video on doing this: https://www.youtube.com/watch?v=12o7WxjAxjM

**Reply by HiDeoo** [reply]
Just watched the video, amazing work, so thorough while still being concise and easy to follow. Loved it, thanks a lot for sharing :raised_hands:

**Reply by delucis** [reply]
Oh wow, yeah! Did a great job condensing my screed above into the simple minimal steps needed 💖
